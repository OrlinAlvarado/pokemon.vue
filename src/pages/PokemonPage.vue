<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>Â¿Quien es este pokemon?</h1>

    <!-- Se puede poner como PokemonPicture o <pokemon-picture></pokemon-picture> -->
    <!-- Las propiedades tambien se pueden poner como pokemon-id en lugar de PokemonId -->
    <PokemonPicture 
      :pokemon-id="pokemon.id" 
      :show-pokemon="showPokemon" />

    <!-- <PokemonOptions 
      :pokemons="pokemonArr"
      @selection="checkAnswer(1, $event)"
    /> -->
    <PokemonOptions 
      :pokemons="pokemonArr"
      @selection="checkAnswer"
    />  
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo Juego
      </button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from '../components/PokemonOptions.vue'
import PokemonPicture from '../components/PokemonPicture.vue'
import getPokemonOptions from '../helpers/getPokemonOptions'

// console.log(getPokemonOptions())
export default {
  components: { PokemonPicture, PokemonOptions },
  data(){
    return {
      pokemonArr: [],
      pokemon: null, 
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray (){
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor( Math.random() * 4 )

      this.pokemon = this.pokemonArr[rndInt]
    },
    checkAnswer(selectedId){
      this.showPokemon = true;
      this.message = selectedId === this.pokemon.id 
                     ? `Adivinate!!!!, es ${ this.pokemon.name }`
                     : `Oops, era ${ this.pokemon.name }` 
      this.showAnswer = true;
    },
    async newGame(){
      this.showAnswer = false;
      this.showPokemon = false;
      this.pokemonArr = []
      this.pokemon = null;
      this.mixPokemonArray()
    }
  },
  mounted(){
    this.mixPokemonArray()
  }
}
</script>
