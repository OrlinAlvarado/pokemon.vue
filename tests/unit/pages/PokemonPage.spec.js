import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage', () => {
    let wrapper
    let mixPokemonArraySpy


    jest.clearAllMocks()

    beforeEach(() => {
        mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount( PokemonPage )
    })


    test('debe de hacer match con el snapshot', () => {
        expect( wrapper.html()).toMatchSnapshot()    
    })

    test('debe de llamar mixPokemonArray al montar', () => {

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })
    

    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

    })
    
    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', async () => {
        const wrapper = await shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        expect( picture.exists() ).toBeTruthy()
        expect( picture.attributes('pokemonid')).toBe('5')

        expect( options.exists() ).toBeTruthy()
        expect( options.attributes('pokemons')).toBeTruthy()

    })

    test('pruebas con checkAnswer', async() => {
        const wrapper = await shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0], 
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(5)

        expect( wrapper.find('h2').exists()).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Adivinate!!!!, es ${ pokemons[0].name }`)

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.message).toBe(`Oops, era ${ pokemons[0].name }`)

    })
    
    
})
