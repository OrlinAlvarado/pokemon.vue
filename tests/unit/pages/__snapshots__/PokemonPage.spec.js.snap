// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage debe de hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>Â¿Quien es este pokemon?</h1><!-- Se puede poner como PokemonPicture o <pokemon-picture></pokemon-picture> -->
  <!-- Las propiedades tambien se pueden poner como pokemon-id en lugar de PokemonId -->
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <!-- <PokemonOptions 
      :pokemons="pokemonArr"
      @selection="checkAnswer(1, $event)"
    /> -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
